### ISSUE TITLE
Not rendering well after resizing screen in web

### ISSUE THREAD

**Issue by jonhygg99** [letsar/flutter_staggered_grid_view#167]
When it resizes the window, it should rebuild and add or remove a widget, but it has a bad render (the widgets stays hidden or it doesn't update).
Here is more information about this issue. A post I did in Stack overflow.
https://stackoverflow.com/questions/66765557/how-can-i-rebuild-if-the-width-of-the-screen-changes-in-web/66767567?noredirect=1#comment118053973_66767567

**Comment by Azzeccagarbugli** [comment]
I can confirm, this is happening to me too!

**Comment by sinhpn92** [comment]
Hummm.
I have same issue in the mobile.
When the layout has changed width, the column do not rebuild, It's has missing column.

**Comment by sinhpn92** [comment]
This is preview for my issue:
[Video from google drive](https://drive.google.com/file/d/1o21OgfppQ7ahkJyhqO6bmD5F2a5qnYKN/view?usp=sharing)

Did you have the same?

**Comment by Azzeccagarbugli** [comment]
Yes, the same kind of issue

**Comment by Harrys76** [comment]
I can get it work by set key param in StaggeredGridView.

e.g:

```
StaggeredGridView.countBuilder(
              key: ValueKey<double>(constraints.maxWidth), <- like this
              physics: const NeverScrollableScrollPhysics(),
              shrinkWrap: true,
              staggeredTileBuilder: (_) => StaggeredTile.fit(1),
              itemCount: _selectionValue.length,
              itemBuilder: (_, index) {
                return _CheckboxItem(_selectionValue[index],
                    _multipleSelectionValue, _isDisabled, widget.state);
              },
              crossAxisCount: grid,
              crossAxisSpacing: 8,
              mainAxisSpacing: 8,
            );
```

**Comment by Azzeccagarbugli** [comment]
The solution posted by @Harrys76 is working correctly! Thanks 🎉

**Comment by sinhpn92** [comment]
Could you explain why it's working? @Harrys76

**Comment by Harrys76** [comment]
> Could you explain why it's working? @Harrys76

it's because the value of key always change while we resize the browser, therefore flutter need to rebuild the widget

**Comment by rydmike** [comment]
I used this solution too, while it works, it is not a very good workaround.

Although instead of tying the key to
`ValueKey<double>(constraints.maxWidth)`

I used on that tied to the `grid` count in the example above. Something like:
`key: ValueKey<String>('HomePage$grid')`

Since the layout issue as far as I could see only happened when the grid (columns) count changed during window resizing.  This worked better since the performance hit is a lot smaller, and the full rebuild only happens when the columns actually changes and the layout breaks.

It still has an issue, other than performance hit, now when it completely throws out the widget and rebuilds it when the key changes,  it then also tosses out the used scroll controllers  scroll position. So when columns changes we are always back at start of view, not nice at all. Tried a few trick to keep it live (like AutomaticKeepAliveClientMixin), but no luck when the key is changed the scroll position is gone too.

My temp hack for this was to listen to the used scroll controller's position, store its value in local state and jump to the position it last had, when dependencies changes and if have a different column count compared what we had before. It kind of works, but even with `scrollController.jumpTo` position it is visible that it jumped, first went to start and then quickly back again to where it was, roughly anyway, scroll position is tricky to keep logical and right with this kind of layout when columns change. Still other than the jump back and forth, it behaves quite OK.

However, it is kind of a hack and performance does still suffer a bit, but nowhere near as much as if using `ValueKey<double>(constraints.maxWidth)` as the key, that was quite poor.

Regarding this layout issue, I have not found a Masonry or Staggered grid like layout package or solution that would be able to handle this layout correctly. 

I gave the WaterfallFlow https://pub.dev/packages/waterfall_flow package a try for the same layout, it has the same issue, but its breaking looks a bit different, but exactly the same use case causes the break, and this same "fix" worked for it too.

I'm beginning to wonder if the issue might be at a lower level layout widget used by eg both StaggeredGridView and WaterfallFlow, or it is just a coincidence. 

cc @letsar

**Comment by soarcreator** [comment]
I am facing this issue right now, so I tried to fix this issue.
But unfortunately, my solution didn't work well.
If here is somebody familiar with how Flutter and Element works, please give me advice or fix this issue in a nicer way.
Thank you.

```
class RenderSliverStaggeredGrid extends RenderSliverVariableSizeBoxAdaptor {
  // added this member variable
  int? lastCrossAxisCount;

  void performLayout() {
    ...
    final configuration = _gridDelegate.getConfiguration(constraints);
  
    // added these lines, but didn't work well
    if (lastCrossAxisCount != null && configuration.crossAxisCount != lastCrossAxisCount) {
      WidgetsBinding.instance?.addPostFrameCallback((dt) {
        (childManager as SliverVariableSizeBoxAdaptorElement).markNeedsBuild();
      });
    }
    lastCrossAxisCount = configuration.crossAxisCount;
    
    ...
  }

```

I hope this issue will be fixed in this original repository soon.

**Comment by letsar** [comment]
Can you test with the latest 0.5.0 preview? https://pub.dev/packages/flutter_staggered_grid_view/versions/0.5.0-dev.1

**Comment by letsar** [comment]
Closing this issue because it's now resolved in 0.5.0-dev.1 🙂
